core/src/main/scala/org/apache/spark/deploy/master/Master.scala

```diff
@@ -78,7 +88,7 @@ private[deploy] class Master(
   private val addressToApp = new HashMap[RpcAddress, ApplicationInfo]
   private val completedApps = new ArrayBuffer[ApplicationInfo]
   private var nextAppNumber = 0
-  private val appIdToUI = new HashMap[String, SparkUI]
+  private val appIdToUI = new ConcurrentHashMap[String, SparkUI]
```

Comment:
we should at least a comment explaining why this needs to be a `ConcurrentHashMap`